[rule based autotagging] Add Delete Rule API Logic

Lindsay-00 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
This PR introduces the Delete Rule API Logic.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ruai0511 committed:
add get rule api logic
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>
rebase

Signed-off-by: Lingxi Chen <lingxich@amazon.com>

ruai0511 committed:
modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

rebase

Signed-off-by: Lingxi Chen <lingxich@amazon.com>

ruai0511 committed:
rebase from main after the schema merged
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

rebase

Signed-off-by: Lingxi Chen <lingxich@amazon.com>

Commit:
delete rule api

rebase

Signed-off-by: Lingxi Chen <lingxich@amazon.com>

Commit:
unit tests

Signed-off-by: Lingxi Chen <lingxich@amazon.com>

kaushalmahi12 commented on plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/action/DeleteRuleResponse.java:
This rule can be replaced with `AcknowledgedResponse`

kaushalmahi12 commented on plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/rule/action/DeleteRuleResponseTests.java:
We will not need this once we remove the source class as commented above to use `AcknowledgedResponse`

kaushalmahi12 commented:
We would also need to add rest based yaml tests once create API pr is merged

chloewqg committed:
Add Unwrapfunction to all the fieldType instaceof check (#17951)

Signed-off-by: Chloe Gao <chloewq@amazon.com>

vibrantvarun committed:
Visit the subquery builder recursively (#17953)

* visit the subquery

Signed-off-by: Varun Jain <varunudr@amazon.com>

* Fix visit method of constant score query and function score query to iterate recursively

Signed-off-by: Varun Jain <varunudr@amazon.com>

---------

Signed-off-by: Varun Jain <varunudr@amazon.com>

kkewwei committed:
Fix loop when updating reposity and creating snapshot (#17532)

Signed-off-by: kkewwei <kewei.11@bytedance.com>
Signed-off-by: kkewwei <kkewwei@163.com>

gargmanik13 committed:
Unset discovery nodes for every transport node actions request (#17682)

* Removed includeDiscoveryNodes from BaseNodesRequest

Signed-off-by: Manik Garg <garmanik@amazon.com>

dependabot[bot] committed:
Bump com.github.spotbugs:spotbugs-annotations from 4.9.0 to 4.9.3 in /test/framework (#17922)

* Bump com.github.spotbugs:spotbugs-annotations in /test/framework

Bumps [com.github.spotbugs:spotbugs-annotations](https://github.com/spotbugs/spotbugs) from 4.9.0 to 4.9.3.
- [Release notes](https://github.com/spotbugs/spotbugs/releases)
- [Changelog](https://github.com/spotbugs/spotbugs/blob/master/CHANGELOG.md)
- [Commits](https://github.com/spotbugs/spotbugs/compare/4.9.0...4.9.3)

---
updated-dependencies:
- dependency-name: com.github.spotbugs:spotbugs-annotations
  dependency-version: 4.9.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update changelog

Signed-off-by: dependabot[bot] <support@github.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump com.microsoft.azure:msal4j from 1.18.0 to 1.20.0 in /plugins/repository-azure (#17925)

* Bump com.microsoft.azure:msal4j in /plugins/repository-azure

Bumps [com.microsoft.azure:msal4j](https://github.com/AzureAD/microsoft-authentication-library-for-java) from 1.18.0 to 1.20.0.
- [Release notes](https://github.com/AzureAD/microsoft-authentication-library-for-java/releases)
- [Changelog](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/changelog.txt)
- [Commits](https://github.com/AzureAD/microsoft-authentication-library-for-java/compare/v1.18.0...1.20.0)

---
updated-dependencies:
- dependency-name: com.microsoft.azure:msal4j
  dependency-version: 1.20.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Updating SHAs

Signed-off-by: dependabot[bot] <support@github.com>

* Update changelog

Signed-off-by: dependabot[bot] <support@github.com>

* Fix test failure by modifying the error message

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Binlong Gao <gbinlong@amazon.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Binlong Gao <gbinlong@amazon.com>

dependabot[bot] committed:
Bump com.google.code.gson:gson from 2.12.1 to 2.13.0 in /test/fixtures/hdfs-fixture (#17923)

* Bump com.google.code.gson:gson in /test/fixtures/hdfs-fixture

Bumps [com.google.code.gson:gson](https://github.com/google/gson) from 2.12.1 to 2.13.0.
- [Release notes](https://github.com/google/gson/releases)
- [Changelog](https://github.com/google/gson/blob/main/CHANGELOG.md)
- [Commits](https://github.com/google/gson/compare/gson-parent-2.12.1...gson-parent-2.13.0)

---
updated-dependencies:
- dependency-name: com.google.code.gson:gson
  dependency-version: 2.13.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update changelog

Signed-off-by: dependabot[bot] <support@github.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: gaobinlong <gbinlong@amazon.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: gaobinlong <gbinlong@amazon.com>

harshavamsi committed:
Fix flaky multimatchquerybuildertests (#18013)

Signed-off-by: Harsha Vamsi Kalluri <harshavamsi096@gmail.com>

10000-ki committed:
Increase default http.max_header_size to 16kb (#18024)

Add change log



Remove duplicate line

Signed-off-by: 10000-ki <10000ki6472@gmail.com>

andrross committed:
Ignore 3.0 branch for breaking changes check (#18030)

We'll need to come back and update this after the 3.0 release but for
now we should not be enforcing breaking changes against 3.0.

Signed-off-by: Andrew Ross <andrross@amazon.com>

cwperks committed:
Limit stack walking in the java agent to frames before the call to AccessController.doPrivileged (#17894)

Signed-off-by: Craig Perkins <cwperx@amazon.com>

andrross committed:
Update JavaVersionChecker to enforce min of JDK21 (#18034)

Signed-off-by: Andrew Ross <andrross@amazon.com>

karenyrx committed:
Remove package org.opensearch.transport.grpc and replace with org.opensearch.plugin.transport.grpc (#18031)

Signed-off-by: Karen Xu <karenxyr@gmail.com>

andrross committed:
Increase Kafka topic polling time in test to 10 seconds (#18040)

We've seen multiple failures where the 3 second timeout was breached.
This happens in the test setup of many different tests so multiple
individual test failures have been reported.

Also replace deprecated method in Kafka API and simplify a couple other
code paths.

Signed-off-by: Andrew Ross <andrross@amazon.com>

peterzhuamazon committed:
Bump OpenSearch Core main branch to 3.0.0 and update key id for native plugin checks (#18039)

* Bump OpenSearch Core main branch to 3.0.0

Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

* Bump CHANGELOG 3.0

Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

* Update release@opensearch.org key id in 3.0 and above

Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

---------

Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

reta committed:
Update bundled JDK to 21.0.7+6 (April 2025 Patch releases) (#18041)

Signed-off-by: Andriy Redko <drreta@gmail.com>

kkewwei committed:
Avoid invalid retries on multiple replicas when querying (#17370)

Signed-off-by: kkewwei <kewei.11@bytedance.com>
Signed-off-by: kkewwei <kkewwei@163.com>

ruai0511 committed:
[Rule based autotagging] Add Get Rule API Logic (#17336)

* add get rule api logic
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* rebase from main after the schema merged
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* extract common logics to libs
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* Add javadocs for libs
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* correct UT
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* modify based on comments
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

r

* refactor code and fix ut

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* remove commented code

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* address comments

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* change method name

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* fix merge conflicts

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* rename queryGroup to workloadGroup

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* add guice binding related changes

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* refactor code to create a generic rule framework structure

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* fix javadoc
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* fix UT
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* restructure tests

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* rebase with mainline

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* fix gradlew file
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* add UT
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

* add action UTs

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* correct the comment

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

* add more UT
Signed-off-by: Ruirui Zhang <mariazrr@amazon.com>

---------

Signed-off-by: Kaushal Kumar <ravi.kaushal97@gmail.com>
Co-authored-by: Kaushal Kumar <ravi.kaushal97@gmail.com>

gaobinlong committed:
Fix ingest pipeline cannot be executed when retry the failed index requests for update_by_query and reindex (#18003)

* Fix ingest pipeline cannot be executed when retry the failed index requests for update_by_query API and reindex API

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

* Modify changelog

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

* Optimize some code

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

* Fix precommit failure

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

---------

Signed-off-by: Binlong Gao <gbinlong@amazon.com>

yupeng9 committed:
[Pull-based ingestion] Emit lag metric for pull-based ingestion poller (#17977)

* emit lag metric for pull-based ingestion poller

Signed-off-by: Yupeng Fu <yupeng@uber.com>

* changelog

Signed-off-by: Yupeng Fu <yupeng@uber.com>

* fix ut

Signed-off-by: Yupeng Fu <yupeng@uber.com>

---------

Signed-off-by: Yupeng Fu <yupeng@uber.com>

skumawat2025 committed:
Add remote to local pool change handling in TargetPoolAllocationDecider (#18027)

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

prudhvigodithi committed:
Skip approximation when `track_total_hits` is set to `true` (#18017)

Signed-off-by: Prudhvi Godithi <pgodithi@amazon.com>

Vikasht34 committed:
Enable concurrent_segment_search auto mode by default (#17978)

* Enable concurrent_segment_search auto mode by default

Signed-off-by: Vikasht34 <viktari@amazon.com>

* Make Default Slice count to 1 for Non-Concurrent Path

Signed-off-by: Vikasht34 <viktari@amazon.com>

* Add tolerance to matrix_stats agg correlation value assertion

The correlation metric could be different for different document distribution across shards, or slices. slice1(doc1,doc2), slice2(doc3,doc4,doc5) could give different correlation from slice1(doc1,doc2,doc3), slice2(doc4,doc5)

The tolerance followed here is 0.000000000000001

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

---------

Signed-off-by: Vikasht34 <viktari@amazon.com>
Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>
Co-authored-by: bowenlan-amzn <bowenlan23@gmail.com>

andrross committed:
Mute ClusterMaxMergesAtOnceIT (#18057)

Signed-off-by: Andrew Ross <andrross@amazon.com>

harshavamsi committed:
Disable approximation when there is terminate after (#18018)

Signed-off-by: Harsha Vamsi Kalluri <harshavamsi096@gmail.com>

harshavamsi committed:
Fix flaky multi match query test (#17935)

Signed-off-by: Harsha Vamsi Kalluri <harshavamsi096@gmail.com>

andrross committed:
Fix race condition in SearchOnlyReplicaIT (#18072)

I observed a failure in local testing due to this issue. The code
previously was assuming it would observe a yellow cluster after
restarting a node, but the node could restart and re-join the cluster
before the assertion causing a failure. The `onNodeStopped` callback
lets the test execute code before the node is restarted, so it should
fix this issue.

Signed-off-by: Andrew Ross <andrross@amazon.com>

gargmanik13 committed:
Enabled default throttling for all tasks submitted to cluster manager (#17711)

Signed-off-by: Manik Garg <gargmanik1317@gmail.com>

rayshrey committed:
Add composite directory factory (#17988)

Signed-off-by: Shreyansh Ray <rayshrey@amazon.com>

reta committed:
Update to Gradle 8.14 (#17942)

Signed-off-by: Andriy Redko <drreta@gmail.com>

Commit:
refactor

Signed-off-by: Lingxi Chen <lingxich@amazon.com>
